### DISCUSSION TITLE
Incremental Static Generation is unusable with multi-process deployments

### DISCUSSION THREAD

**Discussion by xiata** [vercel/next.js#16733] [Help]
# Bug report

Each next-js process uses an application-level cache that doesn't appear to have any synchronization primitives for cases where next-js will be ran as a cluster on the same machine instead of as a single standalone process.

This is very problematic for systems deploying > 1 process to handle requests which is what you'd find typical for a production deployment. Since none of the workers notify each other of revalidation changes, it requires a bare minimum of `n` requests where `n` is equal to the number of workers, before all workers may agree on the same content (assuming no further changes and the generated page is idempotent). Unfortunately most load balancing on cluster managers use a non-deterministic approach to balance traffic, so the actual amount of requests before all workers agree on the content can be significantly higher, especially for large worker pools. 

I've seen upwards of 60 requests on a 24 process deployment before workers appeared to send the same version. For low traffic pages, you'd believe that everyone is seeing whatever new content should be showing, but you could be wrong for a significantly longer period of time than what you set revalidate to.

Below will show the worst possible outcome, a page that changes every generation, showing how chaotic Incremental Static Site Generation can be. The Page auto refreshes at random intervals at less than a second. It will take about 45 seconds before issue shows the clear cache coherency bug.

## To Reproduce

Create this page:

pages/random.js

```javascript
import { useEffect, useState } from 'react';
const random = () => (Math.random() * 1E9 | 0).toString(36);
const duration = age => `${(age / 60) | 0}m ${age % 60}s`;
const revalidate = 30;

export default function ContentTeamHadTooMuchCoffeeToday ({ pid, rand, time }) {
  const [age] = useState(() => ((+new Date() - time) / 1000) | 0);
  useEffect(() => setTimeout(() => global.location.reload(), Math.random() * 1000));
  return <article>
    <h1>{rand} by process {pid}</h1>
    <h2>Age: {duration(age)}. Should revalidate after {duration(revalidate)}</h2>
  </article>;
}

export function getStaticProps () {
  return {
    props: { pid: process.pid, rand: random(), time: +new Date() },
    revalidate,
  };
}
```

Create this pm2 launch file as random.json:
```json
{
  "apps": [{
    "name": "nextjs",
    "script": "node_modules/next/dist/bin/next",
    "args": "start",
    "instances": 8,
    "exec_mode": "cluster"
  }]
}
```

1. next build
2. pm2 start random.json
3. Open http://localhost:3000/random and refresh heavily after 20 seconds.

## Expected behavior

One worker to generate the new content, and other workers become aware of that change so that multi-process deployments don't send different versions of the content that haven't performed their own staleness check yet.

## Workarounds

Don't use Incremental Static Generation with multi-process deployments, defer to cdn. 👎 

## System information

- OS: macOS 10.15.6, CentOS 8
- Any browser
- Version of Next.js: 9.5.2
- Version of Node.js: 12.18.3

## Additional context

One approach that could address this:

Each process should watch the manifest file location for changes, and when change is detected, reload manifest.

When a process is revalidating content
1. Write generated page to cache with content hash in filename.
2. Update manifest and commit to disk.
3. `watch should pick up manifest change across all workers`

Cleaning up old generated content should be fairly easy, such as waiting for some arbitrary amount of time to pass before deleting when the file is not found in the manifest, such as a 2-7 days.

**Comment by xiata** [comment]
Note that undocumented config experimental.sprFlushToDisk only partially fixes the issue-- writing the generated content to disk. When enabled, reloading the next-js workers will appear to start off at one of the generated versions, but as you can guess, the workers are still desynchronized and still take a very long time before all workers serve the same content.

**Comment by xiata** [comment]
I am curious as to why this was moved to a discussion, with no discussion of course, when it's a clear and obvious cache coherence bug.

**Comment by RishikeshDarandale** [comment]
@xiata , Can you please elaborate more on your workaround? I am also have the same concern over `ISR` where next.js server becomes the stateful in multiple process case. I am using AWS ECS fargate tasks as multiple process.

**Comment by lukebarton** [comment]
This is a common problem for anyone trying to use multi-instance deployments. It's rooted in the SSG API. I've been suggesting a rethink on the SSG API for a while now, but it's not my project. I'm going to use this post to make the case once more, and it can be taken or left.

SSG has two features which I'm suggesting are workarounds for a design misstep that Next is very much able to correct:
 - `fallback: true`
 - `revalidate: <n>`

Let's talk about SSG. The SSG API consists of two functions `getStaticProps` and `getStaticPaths`. Defining these functions causes Next to pre-render/generate some pages based on the paths statically, by mimicking calls to these pages, causing the page to render and be cached as part of the app bundle, ready to be served by the application when the application is running.

People said "What if I have a path that wasn't pre-rendered by SSG, currently I get a 404, what can I do?", so Next introduced `fallback: true`. This causes the pages written using the SSG API to render-at-_runtime_ if the cache is missed. Note how similar this is to having a CDN/cache proxy, and hitting a rendered-at-runtime (SSR, `getServerSideProps`) endpoint, if the cache is missed. In fact, the only difference is that SSG causes the cache to be (what we would usually call) pre-warmed. Pre-warmed? Using a standard cache proxy/CDN approach, the first request would miss the (cold) cache, and rendered-at-runtime, then cached by the proxy which is now warm. In the Next SSG API approach, the first request is 'faked' at build time, the file is then bundled, and eventually served by the 'caching' part of Next.js; thus the application starts up with a warm cache, hence is pre-warmed. 

Why is any of that important? The SSG API provides some real benefit to users without a cache. But it's a bespoke API that has limitations, meaning Next.js needs to provide extra 'features' like `fallback: true`. So what, it's already been written, right? Yes. But that's not all, here are some common questions that are also a result of the limitations of this custom SSG API, that require extra features to facilitate:
- "How do I serve a special 404 page with a 404 status code for my (product|blog|whatever) pages that I'm SSG/fallbacking?"
- "How do I (301|302) redirect a user requesting a particular SSG'd page?"
- "How do I read a header from the request when I'm SSG'ing/fallbacking?"
- "How do I set a custom header on a response that I'm SSG'ing/fallbacking?"
- "How do I expire my page after a certain period of time and re-render it?"
- "My caches are out of sync, will Next support some way to sync my caches using shared storage?"

The answer to all of these is that the SSG API doesn't support them. So, Next just needs to do some dev right? Sure, it _could_ do that. But take a step back to when I pointed out how similar the SSG API was to SSR + a cache in front of it, the only difference being the pre-warming. The example suggested having a cache proxy/CDN in front of your app. Next currently provides the caching functionality, and does that with its custom SSG API. I'm suggesting that Next could still provide this caching functionality, instead, moving the caching completely in front of the Next app itself like a traditional cache proxy, and use the SSR `getServerSideProps` API to control things like status codes, headers, etc., and most importantly `cache-control` headers, so that the Next-based cache can respond to the `cache-control` header, just like a standalone cache proxy/CDN would do. 

Providing a conventional, straightforward caching layer using existing caching protocols (`cache-control` is a ubiquitous HTTP caching protocol), this gives the benefits of caching to users without a real standalone cache proxy/CDN, without Next having to provide a custom SSG API, along with features to overcome its limitations. This design could then enable users to turn off this caching layer via configuration if they have their own cache proxy/CDN. This is a very straightforward and easy to understand design, in contrast to "build-time generation that sometimes runs at run-time, but without any control over the expiry of the page, no reading of the request, or control over response headers". 

I suspect the reason that this issue was moved to discussions, and why Next has adopted it's limited SSG API over a more traditional cache proxy design closer to what I've outlined, is that Next is intentionally geared toward deployment on Vercel/now.sh infrastructure, rather than running it yourself in a HA configuration, with your own cache. You only need to look at the issue outlined by the OP to see that the `revalidate: <n>` with caching-as-part-of-the-application functionality is not appropriate in a traditional, non-serverless, server-as-a-process, multi-instance deployment. It only really makes sense if you have a single instance, or use stateless functions with a caching proxy/CDN in front.

So if you've got money (and with some understood caveats), you can get around most of these shortcomings by abandoning the SSG API altogether and instead use the SSR `getServerSideProps` API, in conjunction with your own cache proxy/CDN. Except that some of the more interesting features that Next offers, like data prefetching, require pages to use the SSG API. But if I use the SSG API, it comes with all the limitations I've already mentioned. The reason provided to answer why you can't do data prefetching with `getServerSideProps` is that "you should only prefetch cached stuff because server load". If I've brought my own cache, that isn't an issue anymore.

The custom caching API has a bunch of shortcomings, there are different limitations which need additional work to be addressed. I've proposed a rethink on the SSG API and I'm optimistic that someone will take it seriously.

/cc @timneutkens

**Comment by christophersjchow** [comment]
Doesn't look like there's much movement in this area. 

We've run into this issue on a large e-commerce site too with the mitigation largely short revalidate TTLs. 

(Side note there are a number of bugs we hope to contribute back around how Next handles cache headers that we've worked around by inserting an nginx reverse proxy). 

We have also noticed a subtle challenge with ISR and Next's handling of preloading. In a multi process or server setup when deploying a new version of the application there is a chance the props change as old servers are rotated out and new servers are rotated in (in our case a Kube pod). 

This means if the shape of your props change you may have errors due to prefetching of `_next/data`. This is effectively the same as a forward compatible schema migration in a database where it takes 3 deployments to remove a field from props to ensure error free deployments. Just another gotcha with static rendering.

**Comment by shahinsafarali** [comment]
This snippet might solve the issue:

https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#self-hosting-isr

```Javascript
module.exports = {
  experimental: {
    // Defaults to 50MB
    isrMemoryCacheSize: 0,
  },
}
```
